tests/e2e/common/diagnostics/kubernetes.go

```diff
@@ -0,0 +1,161 @@
+/**
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+
+package diagnostics
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+)
+
+type nodes struct {
+	*Config
+}
+
+type namespaces struct {
+	*Config
+}
+
+type pods struct {
+	*Config
+}
+
+type deployments struct {
+	*Config
+}
+
+type daemonsets struct {
+	*Config
+}
+
+type jobs struct {
+	*Config
+}
+
+func (c nodes) Collect(ctx context.Context) error {
+	nodes, err := c.Clientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Error getting nodes")
+		return err
+	}
+
+	if err := c.outputTo("nodes.yaml", nodes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c namespaces) Collect(ctx context.Context) error {
+	namespaces, err := c.Clientset.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Error getting namespaces")
+		return err
+	}
+
+	if err := c.outputTo("namespaces.yaml", namespaces); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c daemonsets) Collect(ctx context.Context) error {
+	daemonsets, err := c.Clientset.AppsV1().DaemonSets(c.namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Error getting daemonsets")
+		return err
+	}
+
+	if err := c.outputTo("daemonsets.yaml", daemonsets); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c deployments) Collect(ctx context.Context) error {
+	deployments, err := c.Clientset.AppsV1().Deployments(c.namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Error getting deployments")
+		return err
+	}
+
+	if err := c.outputTo("deployments.yaml", deployments); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c pods) Collect(ctx context.Context) error {
+	pods, err := c.Config.Clientset.CoreV1().Pods(c.namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Error getting pods")
+		return err
+	}
+
+	if err := c.outputTo("pods.yaml", pods); err != nil {
+		return err
+	}
+
+	// Get all logs per pod
+	for _, pod := range pods.Items {
+		podLogFile, err := c.createFile(fmt.Sprintf("%s_logs.log", pod.Name))
+		if err != nil {
+			return err
+		}
+		defer podLogFile.Close()
```

Comment:
Question: Is this executed on each loop iteration?

Reply:
each loop iteration is a different pod, and will be saved to a diff file.

Reply:
Yes, I understand. My question was whether the defer statement is triggered in each loop.

Looking at https://go.dev/play/p/ggT4Sim9X1u, this is not the case. It seems as if we would first create all files before unrolling the defers -- that is to say that the loop body doesn't represent a conext that is exited.

The simplest solution to ensure that the pod log is created and closed in the same iteration is to call a function in the loop body instead. See https://go.dev/play/p/vuQndJk6B0U

We could do the following to ensure that we are closing these files properly:
```
diff --git a/tests/e2e/common/diagnostics/kubernetes.go b/tests/e2e/common/diagnostics/kubernetes.go
index 637a95ac..4dd88d4a 100644
--- a/tests/e2e/common/diagnostics/kubernetes.go
+++ b/tests/e2e/common/diagnostics/kubernetes.go
@@ -19,11 +19,11 @@ package diagnostics
 import (
 	"bufio"
 	"context"
+	"errors"
 	"fmt"
 
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/klog/v2"
 )
 
 type nodes struct {
@@ -112,32 +112,50 @@ func (c pods) Collect(ctx context.Context) error {
 		return err
 	}
 
-	// Get all logs per pod
+	var podLogCollectors []Collector
 	for _, pod := range pods.Items {
-		podLogFile, err := c.createFile(fmt.Sprintf("%s.log", pod.Name))
-		if err != nil {
-			return err
+		plc := podLogCollector{
+			Config: c.Config,
+			name:   pod.Name,
 		}
-		defer podLogFile.Close()
+		podLogCollectors = append(podLogCollectors, plc)
+	}
 
-		req := c.Clientset.CoreV1().Pods(c.namespace).GetLogs(pod.Name, &v1.PodLogOptions{})
-		podLogs, err := req.Stream(ctx)
-		if err != nil {
-			klog.ErrorS(err, "error getting pod logs")
-			continue
-		}
+	var errs error
+	for _, collector := range podLogCollectors {
+		errs = errors.Join(errs, collector.Collect(ctx))
+	}
 
-		buf := bufio.NewScanner(podLogs)
-		for buf.Scan() {
-			if _, err := podLogFile.Write(buf.Bytes()); err != nil {
-				klog.ErrorS(err, "error getting pod logs", "podName", pod.Name)
-			}
-		}
-		if err := buf.Err(); err != nil {
-			klog.ErrorS(err, "error reading pod log")
-		}
+	return errs
+}
+
+type podLogCollector struct {
+	*Config
+	name string
+}
+
+func (c podLogCollector) Collect(ctx context.Context) error {
+	podLogFile, err := c.createFile(fmt.Sprintf("%s.log", c.name))
+	if err != nil {
+		return err
 	}
+	defer podLogFile.Close()
+
+	req := c.Clientset.CoreV1().Pods(c.namespace).GetLogs(c.name, &v1.PodLogOptions{})
+	podLogs, err := req.Stream(ctx)
+	if err != nil {
+		return fmt.Errorf("error getting logs for pod %v: %w", c.name, err)
 
+	}
+	buf := bufio.NewScanner(podLogs)
+	for buf.Scan() {
+		if _, err := podLogFile.Write(buf.Bytes()); err != nil {
+			return fmt.Errorf("error writing logs for pod %v: %w", c.name, err)
+		}
+	}
+	if err := buf.Err(); err != nil {
+		return fmt.Errorf("error reading logs for pod %v: %w", c.name, err)
+	}
 	return nil
 }
```

Reply:
Thanks for the suggestion, I implemented it, with a small mod